### PR TITLE
e2e test: core scenario for beta

### PR DESCRIPTION
**Pull request by kadel**
The new clean e2e scenario, that tries to simulate basic flow.
This can be a replacement for the "old" template.

This will be required to pass before we release beta version.

Watch test is missing, I'll add that in another PR.

### CODE REVIEW

**Comment by cdrage** [comment]
`template_cleantest_test.go` was deleted, is this normal?

I saw you create the new "flow", should we have another template? 

I'm wondering this since I'm currently implementing an e2e test for PR https://github.com/openshift/odo/pull/1446

**Comment by kadel** [comment]
> `template_cleantest_test.go` was deleted, is this normal?

yes, to be honest, there was some mistakes in that template that could lead people to a wrong path :-(
I think that having a concrete  useful working test that can be used as a example is better than "half backed" template that doesn't  test anything properly

> 
> I saw you create the new "flow", should we have another template?

why another template?

> 
> I'm wondering this since I'm currently implementing an e2e test for PR #1446

This should be independent of any PRs. It is just generic testing scenario.

**Comment by kadel** [comment]
/hold cancel
ping @amitkrout  @mohammedzee1000

**Comment by amitkrout** [comment]
I will take a look and verify

**Comment by surajnarwade** [comment]
nodejs flake, restarted the travis

**Comment by kadel** [comment]
ping @surajnarwade @amitkrout all test green

**Comment by kadel** [comment]
> It’s a good idea to keep e2e scenario test file in a separate dir, however I can see more improvement can be done to avoid duplicate helper/source file in tests dir. This improvement can be done later after beta release in a proper way.

My idea was to start with separating scenarios (real e2e test) and our current e2e tests (more component level/integration tests)

I thought that in future we would rename `tests/e2e` to `tests/integration` and `tests/e2escenarios`  to `tests/e2e`.  But it is something for longer discussion it is clear that we need a better way to organize tests, this is just one of the options before we figure out the right way

**Comment by amitkrout** [comment]
@girishramnani @mohammedzee1000 Can please add ```lgtm``` if looks good to you  review 

> > It’s a good idea to keep e2e scenario test file in a separate dir, however I can see more improvement can be done to avoid duplicate helper/source file in tests dir. This improvement can be done later after beta release in a proper way.
> 
> My idea was to start with separating scenarios (real e2e test) and our current e2e tests (more component level/integration tests)
> 
> I thought that in future we would rename `tests/e2e` to `tests/integration` and `tests/e2escenarios` to `tests/e2e`. But it is something for longer discussion it is clear that we need a better way to organize tests, this is just one of the options before we figure out the right way

Right, I am also thinking the same way. For example we can follow something like 
```
Amits-MacBook-Pro:tests amit$ tree -L 1
.
├── OWNERS
├── e2e-scenario
├── helper
├── integration
├── test-src
└── test-template

5 directories, 1 file
Amits-MacBook-Pro:tests amit$
```

**Comment by amitkrout** [comment]
> @girishramnani @mohammedzee1000 Can please add `lgtm` if looks good to you review
> 
> > > It’s a good idea to keep e2e scenario test file in a separate dir, however I can see more improvement can be done to avoid duplicate helper/source file in tests dir. This improvement can be done later after beta release in a proper way.
> > 
> > 
> > My idea was to start with separating scenarios (real e2e test) and our current e2e tests (more component level/integration tests)
> > I thought that in future we would rename `tests/e2e` to `tests/integration` and `tests/e2escenarios` to `tests/e2e`. But it is something for longer discussion it is clear that we need a better way to organize tests, this is just one of the options before we figure out the right way
> 
> Right, I am also thinking the same way. For example we can follow something like
> 
> ```
> Amits-MacBook-Pro:tests amit$ tree -L 1
> .
> ├── OWNERS
> ├── e2e-scenario
> ├── helper
> ├── integration
> ├── test-src
> └── test-template
> 
> 5 directories, 1 file
> Amits-MacBook-Pro:tests amit$
> ```

Created #1581

**Comment by kadel** [comment]
> @kadel Unfortunately i see failures locally :(
> 
> Seems `make test-e2e-scenarios` is not a part of travis and prow CI run. Can you please add the test to both the CI

yes, it the test will fail as it is testing stuff that were not fixed yet.

Can't add it to CI until this is merged.

**Comment by amitkrout** [comment]
> > @kadel Unfortunately i see failures locally :(
> > Seems `make test-e2e-scenarios` is not a part of travis and prow CI run. Can you please add the test to both the CI
> 
> yes, it the test will fail as it is testing stuff that were not fixed yet.
> 
> Can't add it to CI until this is merged.

Ahh...Thank for the info. 
I would say let wait for all the pr for beta is merged first then you can enable this in CI.

File: tests/e2escenarios/core_beta_flow.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+			JustAfterEach(func() {
+				oc.SwitchProject(originalProject)
+			})
+
+			It("odo component should fail if there already is .odo dir", func() {
+				helper.CmdShouldPass("odo", "component", "create", "nodejs", "--context", context)
+				helper.CmdShouldFail("odo", "component", "create", "nodejs", "--context", context)
+			})
+
+			It("odo config should fail if there is no .odo dir", func() {
```
Comment (surajnarwade) [inline tests/e2escenarios/core_beta_flow.go]: this test should be put before previous test because previous test will create `.odo` dir

**Comment by amitkrout** [comment]
> > I would say let wait for all the pr for beta is merged first then you can enable this in CI.
> 
> Why? How are we going to know when this is passing if it is not run?
> Would it be a problem if this is running before everything is in?
> Also merging this is not going to affect anything as the new job needs to be added in openshift/release

My assumption is based on local test failure and the failure reason is as you explained above https://github.com/openshift/odo/pull/1559#issuecomment-481212494. So I thought if the e2e for beta is failing then there is no reason to merge it to master.

From the above set of question I can assume your thought behind this 
```
1.  Let e2e test for beta ready in the master
2. Once the all the required pr for beta is in then just enable the e2e test job 
3. Then e2e test will test and verify the beta bits for release
```

Hope my guess is right ;)

Considering your assumption I am approving the pr
/approve

File: tests/e2escenarios/core_beta_flow.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+			JustAfterEach(func() {
+				helper.Chdir(originalDir)
+			})
+
+			It("'odo component' should fail if there already is .odo dir", func() {
+				helper.CmdShouldPass("odo", "component", "create", "nodejs")
+				helper.CmdShouldFail("odo", "component", "create", "nodejs")
+			})
+
+			It("'odo config' should fail if there is no .odo dir", func() {
```
Comment (surajnarwade) [inline tests/e2escenarios/core_beta_flow.go]: this should be put before previous test

Comment (surajnarwade) [reply tests/e2escenarios/core_beta_flow.go]: ideally command will run successfully because previous test will create `.odo`

**Comment by girishramnani** [comment]
\lgtm

**Comment by surajnarwade** [comment]
@kadel , ignore the review, I was bit confused but CI seems to be :green_heart:

**Comment by kadel** [comment]
> > > I would say let wait for all the pr for beta is merged first then you can enable this in CI.
> > 
> > 
> > Why? How are we going to know when this is passing if it is not run?
> > Would it be a problem if this is running before everything is in?
> > Also merging this is not going to affect anything as the new job needs to be added in openshift/release
> 
> My assumption is based on local test failure and the failure reason is as you explained above [#1559 (comment)](https://github.com/openshift/odo/pull/1559#issuecomment-481212494). So I thought if the e2e for beta is failing then there is no reason to merge it to master.
> 
> From the above set of question I can assume your thought behind this
> 
> ```
> 1.  Let e2e test for beta ready in the master
> 2. Once the all the required pr for beta is in then just enable the e2e test job 
> 3. Then e2e test will test and verify the beta bits for release
> ```
> Hope my guess is right ;)
> 
> Considering your assumption I am approving the pr
> /approve

yes, sorry if it was not clear. I thought that we discussed this at the blj call that we have when I opened this.

File: tests/e2escenarios/core_beta_flow.go (hunk 2 above)
Comment (kadel) [reply tests/e2escenarios/core_beta_flow.go]: Each `It` is an independent test case. There can't be any dependencies on the order in which the `It` sections are run.

**Pull request merged**